diff --git a/vllm_ascend/models/deepseek_v4.py b/vllm_ascend/models/deepseek_v4.py
old mode 100644
new mode 100755
index 231f07b8..22c159ae
--- a/vllm_ascend/models/deepseek_v4.py
+++ b/vllm_ascend/models/deepseek_v4.py
@@ -1155,11 +1155,14 @@ class AscendDeepseekV4ForCausalLM(nn.Module, SupportsPP,
 
             if "sink" in name:
                 # Handle attention sinks (distributed across ranks)
-                param = params_dict[name]
-                narrow_weight = loaded_weight.narrow(0, head_start,
-                                                     heads_per_rank)
-                param.data.copy_(narrow_weight)
-                loaded_params.add(name)
+                # skip error
+                try:
+                    param = params_dict[name]
+                    narrow_weight = loaded_weight.narrow(0, head_start, heads_per_rank)
+                    param.data.copy_(narrow_weight)
+                    loaded_params.add(name)
+                except Exception as e:
+                    print(f"{name} 发生错误: {e}")
                 continue
 
             is_fusion_moe_shared_experts_layer = (
@@ -1196,10 +1199,13 @@ class AscendDeepseekV4ForCausalLM(nn.Module, SupportsPP,
 
                 if is_pp_missing_parameter(name, self):
                     continue
-
-                param = params_dict[name]
-                weight_loader = param.weight_loader
-                weight_loader(param, loaded_weight, shard_id)
+                # skip error
+                try:
+                    param = params_dict[name]
+                    weight_loader = param.weight_loader
+                    weight_loader(param, loaded_weight, shard_id)
+                except Exception as e:
+                    print(f"{name} 发生错误: {e}")
                 break
             else:
                 is_expert_weight = False
@@ -1265,27 +1271,32 @@ class AscendDeepseekV4ForCausalLM(nn.Module, SupportsPP,
 
                         if is_pp_missing_parameter(name_mapped, self):
                             continue
-
-                        param = params_dict[name_mapped]
-                        # We should ask the weight loader to return success or
-                        # not here since otherwise we may skip experts with
-                        # other available replicas.
-                        weight_loader = typing.cast(Callable[..., bool],
-                                                    param.weight_loader)
-                        success = weight_loader(
-                            param,
-                            weight_to_load,
-                            name_mapped,
-                            shard_id=shard_id,
-                            expert_id=expert_id,
-                            return_success=True,
-                        )
-                        if success:
-                            if not is_fusion_moe_shared_experts_layer:
-                                name = name_mapped
-                            else:
-                                loaded_params.add(name_mapped)
-                            break
+                        
+                        # skip error
+                        try:
+                            param = params_dict[name_mapped]
+                            # We should ask the weight loader to return success or
+                            # not here since otherwise we may skip experts with
+                            # other available replicas.
+                            weight_loader = typing.cast(
+                                Callable[..., bool], param.weight_loader
+                            )
+                            success = weight_loader(
+                                param,
+                                weight_to_load,
+                                name_mapped,
+                                shard_id=shard_id,
+                                expert_id=expert_id,
+                                return_success=True,
+                            )
+                            if success:
+                                if not is_fusion_moe_shared_experts_layer:
+                                    name = name_mapped
+                                else:
+                                    loaded_params.add(name_mapped)
+                                break
+                        except Exception as e:
+                            print(f"{name} 发生错误: {e}")
                     else:
                         if is_expert_weight:
                             # We've checked that this is an expert weight
@@ -1304,11 +1315,24 @@ class AscendDeepseekV4ForCausalLM(nn.Module, SupportsPP,
 
                         if is_pp_missing_parameter(name, self):
                             continue
-
-                        param = params_dict[name]
-                        weight_loader = getattr(param, "weight_loader",
-                                                default_weight_loader)
-                        weight_loader(param, loaded_weight)
+                        
+                        # skip error
+                        try:
+                            param = params_dict[name]
+
+                            if "wo_a" in name and len(param.shape) == 3:
+                                n_local_groups = param.shape[0]
+                                o_lora_rank = param.shape[2]
+                                hidden = param.shape[1]
+                                param.data = param.data.transpose(2, 1).contiguous().view(n_local_groups * o_lora_rank,  hidden)
+
+                            weight_loader = getattr(
+                                param, "weight_loader", default_weight_loader
+                            )
+                            weight_loader(param, loaded_weight)
+
+                        except Exception as e:
+                            logger.warning(f"Error in load {name}: {e}")
             if not is_fusion_moe_shared_experts_layer:
                 loaded_params.add(name)
 
